design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,542 @@
+xlang_fail if there is no equivalent mapping. After translating the error, the projection should
+call xlang's error origination API, `xlang_originate_error`, to create the xlang error info containing the
+error and the diagnostic information collected by the language. The created error info will be returned
+and the projection should take ownership of it. Note that in situations where memory allocations fail,
+the origination API may choose to return a statically allocated error info with only the `xlang_result`.
+
+xlang_originate_error will be defined in the PAL and will take the following parameters:
+- a `xlang_result` with the translated xlang error for the language specific error that occurred
+- a `xlang_string` with the error message
+- a `xlang_string` with an identifier for the projection
+- a `xlang_string` with the language specific error that occurred
+- a `IXlangErrorExecutionTrace` with the execution trace if available
+- an `IUnknown` pointer with any additional language specific information (can be null)
+
+The `xlang_result` parameter is the only required parameter for the creation of the error info, but the
+other parameters are recommended, if available, to assist the developer with diagnosing the error.
+ 
+```cpp
+[[nodiscard]] IXlangErrorInfo* xlang_originate_error(
+    xlang_result error,
+    xlang_string message,
+    xlang_string projectionIdentifier,
+    xlang_string languageError,
+    IXlangErrorExecutionTrace* executionTrace,
+    IUnknown* languageInformation)
+```
+
+If the language error is a propagation of a previous xlang error (i.e. the component called another xlang
+projected component which returned an error), then the associated error info should be retrieved.
+The retrieved xlang error info and the associated xlang error should be used as the error instead of
+translating the language error and creating a new xlang error info. This will allow the identity
+of the error to be independent of the translations it may go through as it propagates through
+various languages and for it to be defined as what the initial projection which returned the error
+determined it was equivalent to. Diagnostic information collected by the language should still be
+added to the error info as it can provide useful information in the case the error becomes a failure.
+This can be done by a call to `PropagateError` which will add the provided information to the
+error info.
+
+`PropagateError` will be defined in the `IXlangErrorInfo` interface and will take the following parameters:
+- a `xlang_string` indicating an identifier for the language
+- a `xlang_string` with the language specific error
+- a `IXlangErrorExecutionTrace` with the execution trace if available
+- an `IUnknown` pointer with any additional language specific information (can be null)
+
+```cpp
+void PropagateError(
+    xlang_string projectionIdentifier,
+    xlang_string languageError,
+    IXlangErrorExecutionTrace* executionTrace,
```

Comment:
executionTrace should be IUnknown, to permit both:
1. simple IStringable implementation
2. versioning later (any interface would, but IUnknown makes it clearer)

Reply:
Addressed.